Core/include/Acts/Detector/detail/CylindricalDetectorHelper.hpp

```diff
@@ -166,15 +168,82 @@ DetectorComponent::PortalContainer wrapInZR(
 /// @brief Helper method to extract r,z,phi boundaries for
 /// eventual grid volume search
 ///
+/// @tparam volume_container_t the type of the container
+///
 /// @param gctx the geometry context of the call
 /// @param volumes the volumes at input
 /// @param logLevel is the screen logging level
 ///
 /// @return extracted boundary values
+template <typename volume_container_t>
 std::array<std::vector<ActsScalar>, 3u> rzphiBoundaries(
-    const GeometryContext& gctx,
-    const std::vector<const DetectorVolume*>& volumes,
-    Acts::Logging::Level logLevel = Acts::Logging::INFO);
+    const GeometryContext& gctx, const volume_container_t& volumes,
+    Acts::Logging::Level logLevel = Acts::Logging::INFO) {
+  // The local logger
+  ACTS_LOCAL_LOGGER(getDefaultLogger("CylindricalDetectorHelper", logLevel));
+
+  ACTS_DEBUG("Estimate R/z/phi boundaries of  " << volumes.size()
+                                                << " volumes.");
+
+  // The return boundaries
+  std::array<std::vector<ActsScalar>, 3u> boundaries;
+
+  // The map for collecting
+  std::array<std::map<ActsScalar, size_t>, 3u> valueMaps;
+  auto& rMap = valueMaps[0u];
+  auto& zMap = valueMaps[1u];
+  auto& phiMap = valueMaps[2u];
+
+  auto fillMap = [&](std::map<ActsScalar, size_t>& map,
+                     const std::array<ActsScalar, 2u>& values) {
+    for (auto v : values) {
+      if (map.find(v) != map.end()) {
+        ++map[v];
+      } else {
+        map[v] = 1u;
+      }
+    }
+  };
+
+  // Loop over the volumes and collect boundaries
+  for (const auto& v : volumes) {
+    if (v->volumeBounds().type() == VolumeBounds::BoundsType::eCylinder) {
+      auto bValues = v->volumeBounds().values();
+      // The min/max values
+      ActsScalar rMin = bValues[CylinderVolumeBounds::BoundValues::eMinR];
+      ActsScalar rMax = bValues[CylinderVolumeBounds::BoundValues::eMaxR];
+      ActsScalar zCenter = v->transform(gctx).translation().z();
+      ActsScalar zHalfLength =
+          bValues[CylinderVolumeBounds::BoundValues::eHalfLengthZ];
+      ActsScalar zMin = zCenter - zHalfLength;
+      ActsScalar zMax = zCenter + zHalfLength;
+      ActsScalar phiCenter =
+          bValues[CylinderVolumeBounds::BoundValues::eAveragePhi];
+      ActsScalar phiSector =
+          bValues[CylinderVolumeBounds::BoundValues::eHalfPhiSector];
+      ActsScalar phiMin = phiCenter - phiSector;
+      ActsScalar phiMax = phiCenter + phiSector;
+      // Fill the maps
+      fillMap(rMap, {rMin, rMax});
+      fillMap(zMap, {zMin, zMax});
+      fillMap(phiMap, {phiMin, phiMax});
+    }
+  }
+
+  for (auto [im, map] : enumerate(valueMaps)) {
+    for (auto [key, value] : map) {
```

Comment:
If the ```value``` is not needed anywhere, can't you fill the boundaries into the vector directly?

Reply:
Absolutely right, I think I wanted to add some statistics, but who needs to know that. Changed.